### ISSUE TITLE
If statement does not narrow types properly if the condition is extracted to a variable

### ISSUE THREAD

**Issue by rpmrmartin** [microsoft/pyright#2386]
**Describe the bug**
Normally, inside an if statement like `if optional is not None: ...`, pyright knows that `optional` is not `None`. However, if the condition is assigned to a variable just before the if statement, then the variable is used as the condition of the if statement, pyright fails to recognize that `optional` is not `None`.

**To Reproduce**
```python
from typing import Optional

def works(optional: Optional[str]) -> None:
	if optional is not None:
		optional.capitalize()

def doesntWork(optional: Optional[str]) -> None:
	condition = optional is not None
	if condition:
		optional.capitalize() # "capitalize" is not a known member of "None"
```

**Expected behavior**
Pyright recognizes that `optional` is not `None` inside the if statement

**VS Code extension or command-line**
VS Code extension (Pylance v2021.9.4)
Command line v1.1.175

**Comment by erictraut** [comment]
I figured out how to implement this in a way that imposes no significant performance overhead. There are some limitations, but it covers the common cases. For details and examples, refer to https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#narrowing-based-on-local-variable.

**Comment by erictraut** [comment]
This is included in pyright 1.1.195, which I just published. It will also be included in the next release of pylance.
